peps/pep-0735.rst

```diff
@@ -0,0 +1,476 @@
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+Tools which support Dependency Groups MUST support both string and object
+representations of requirements.
+
+When unrecognized keys are encountered in requirement specifiers or the
+``requirements`` table, tools MUST NOT fail. They MAY emit warnings.
+
+When installing the current package (``.``) from a dependency group, tools
+SHOULD prefer editable installs over non-editable installs. They MAY provide
+users with options to configure this behavior.
+
+Use of Dependency Groups
+------------------------
+
+Tools which support Dependency Groups are expected to provide new options and
+interfaces to allow users to install from Dependency Groups. Implementations
+may wish to treat them similarly to their current treatments of
+``requirements.txt`` files, or more similarly to extras.
+
+No syntax is defined for expressing the dependency group of a package, for two
+reasons:
+
+* it would not be valid to refer to the dependency groups of a third-party
+  package from pypi (because the data is defined to be unpublished)
+
+* there is not guaranteed to be a current package for dependency groups -- part
+  of their purpose is to support non-package projects
+
+For example, a possible ``pip`` interface for installing dependency groups
+would be:
+
+.. code:: shell
+
+    pip install --dependency-groups=test,typing
+
+Note that this is only an example. This PEP does not declare any requirements
+for how tools support the installation of Dependency Groups.
+
+Reference Implementation
+========================
+
+TODO! STUB!
+
+Planned reference implementation:
+A very simple "environment manager" which can be used to build virtualenvs from
+specified Dependency Groups. It will therefore need to fully support parsing of
+Dependency Groups.
+
+Backwards Compatibility
+=======================
+
+At time of writing, the ``requirements`` table is reserved for use by PEPs,
+meaning that this lays claim to a previously unused namespace.
+There should therefore be no direct backwards compatibility concerns.
+
+Future Compatiiblity
+--------------------
+
+This PEP defines explicit behaviors for tools when encountering unknown keys
+which must be followed in order to remain spec-compliant. These behaviors are
+meant to ensure that future PEPs can extend the data with very clear
+expectations about how existing tools will behave.
```

Comment:
But that could also allow for surprise "success" when unexpected things are encountered. Compare that to knowing whether your tool supports a newer standard based on whether something succeeded or failed (or at least warned you).

Reply:
I'm not sure what the best course to chart here is. We could leave it up to the tools to decide how they want to handle unknown keys and only go so far as to say that they "SHOULD warn" on unrecognized keys. It's weaker as a specification, but maybe that's for the best in this case?

Reply:
It's ultimately up to you as the PEP author which one you think will be the best outcome while getting the PEP accepted.

Reply:
I looked a lot at how PEP 621 phrased the keys and values in its defined tables and tried to follow that model, which is generally to be a bit more explicit about what's a requirement and not to say anything unnecessary about unspecified fields. :crossed_fingers: that I got the balance right.